nix-bundle.sh

```diff
@@ -1,56 +1,165 @@
+}
 
-Create a single-file bundle from the nixpkgs attribute "TARGET".
-EXECUTABLE should be relative to the TARGET's output path.
+check_arg() {
+    if [ $# -le 1 ]
+    then
+        echo "error: flag '$1' requires an argument" >&2
+        fail
+    fi
+}
 
-For example:
+nix_file=`dirname $0`/default.nix
+
+# Read arguments
+
+build_flags=()
+unset bin
+targets=()
+
+while [ $# -ne 0 ]
+do
+    case "$1" in
+        -f|--file|-I|--include)
+            check_arg "$@"
+            build_flags+=("$1")
+            shift
+            build_flags+=("$1")
+            shift
+            ;;
+        --arg|--argstr)
+            if [ $# -le 2 ]
+            then
+                echo "error: flag '$1' requires two arguments" >&2
+                fail
+            fi
+            build_flags+=("$1")
+            shift
+            build_flags+=("$1")
+            shift
+            build_flags+=("$1")
+            shift
+            ;;
+        --bin)
+            check_arg "$@"
+            shift
+            bin="$1"
+            shift
+            ;;
+        -h|--help)
+            cat <<EOF >&2
+Usage: $0 <FLAGS>... [<TARGET>] <targets>...
+
+Summary: Similar to nix build, but creates a single-file bundle.
 
-$ $0 hello /bin/hello
-$ ./hello
-Hello, world!
+Flags:
+      --arg <NAME> <EXPR>       argument to be passed to Nix functions
+      --argstr <NAME> <STRING>  string-valued argument to be passed to Nix functions
+  -f, --file <FILE>             evaluate FILE rather than the default
+  -I, --include <PATH>          add a path to the list of locations used to look up <...> file names
+      --bin <BINARY>            the binary to bundle, relative to TARGET's output path; defaults to
+                                the lexicographically first file inside /bin
 
+Examples:
+
+  To bundle GNU Hello from NixOS 19.03:
+  $ $0 -f channel:nixos-19.03 hello --bin /bin/hello
+
+  To bundle the executable from within a project that has a default.nix for build:
+  $ $0 --bin /bin/name-of-the-executable
 EOF
+            exit 0
+            ;;
+        --)
+            shift
+            break
+            ;;
+        -*)
+            echo "error: unrecognized flag '$1'" >&2
+            fail
+            ;;
+        *)
+            targets+=("$1")
+            shift
+            ;;
+    esac
+done
 
-    exit 1
-fi
+while [ $# -ne 0 ]
+do
+    targets+=("$1")
+    shift
+done
 
-nix_file=`dirname $0`/default.nix
+# Create tmp directory for the links produced by `nix build`
 
-target="$1"
-shift
+linkdir=$(mktemp -d)
+build_flags+=("--out-link")
+build_flags+=("$linkdir/result")
 
-extraTargets=
-if [ "$#" -gt 1 ]; then
-    while [ "$#" -gt 1 ]; do
-        extraTargets="$extraTargets $1"
-        shift
-    done
+# Build targets
+
+target_path=
+extra_target_paths=
+
+for target in "${targets[@]}"
+do
+    nix build "${build_flags[@]}" "$target" || exit $?
+    if [ -z "$target_path" ]
+    then
+        target_path="$(readlink "$linkdir/result")"
```

Comment:
How does this handle multiple targets?

Reply:
extra_target_paths contains the store paths (separated by whitespace) to the extra targets.